### PR TITLE
Using the FIWARE logging format and the 'Unica-Correlator' HTTP header

### PR DESCRIPTION
**Pull request by gtorodelvalle**
- Implements https://github.com/telefonicaid/IoT-STH/issues/3

We use https://www.npmjs.com/package/winston as an asynchronous logging system for Node.js.

The current implementation offers distinct parameters or environment variables which makes it possible to configure the logging behaviour when starting the STH server. This environment variables are:
- LOG_LEVEL: The level of the messages to be logged. According to the FIWARE logging guidelines, these levels are: trace, info, debug, warn, error, fatal.
- LOG_TO_CONSOLE: If the value passed is distinct than 'false', the messages will be logged to the console.
- LOG_TO_FILE: If the value passed is distinct than 'false', the messages will be logged to a file.
- LOG_DIR: The path to the directory where the file for the logging will be created if it does not exist.
- LOG_FILE_NAME: The name of the file where the logs will be stored.

This pull request also includes the management of the 'Unica-Correlator' HTTP header. In case the request includes this header, it is included in the response. The Unica-Correlator value is also used in the logs to track the requests made to the server.

### CODE REVIEW

File: src/sth_app.js [hunk 1]
```diff
@@ -24,7 +24,9 @@
     }
 
     if (err) {
-      sthLogger.error(err);
+      sthLogger.fatal(err.toString(), {
+        operationType: sthConfig.OPERATION_TYPE.SHUTDOWN
```
Comment (frbattid) [inline src/sth_app.js]: Just to understand it, the `operationType` is passed as useful information for the logging process, right? I mean, in this particular case something like "a fatal error occured when trying ti shutdown the server" is logged.

File: src/sth_app.js [hunk 2]
```diff
@@ -39,22 +41,37 @@
...
         });
     }
   );
 
+  // In case Control+C is clicked, exit gracefully
   process.on('SIGINT', function () {
     return exitGracefully(null);
   });
+
+  // In case of an uncaught exception exis gracefully
```
Comment (frbattid) [inline src/sth_app.js]: Nice!

NTC

File: src/sth_configuration.js [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
+  var path = require('path');
+
   var ENV = process.env;
 
   var dbUsername = ENV.DB_USERNAME || '';
   var dbPassword = ENV.DB_PASSWORD || '';
 
   module.exports = {
+    LOG_LEVEL: ENV.LOG_LEVEL || 'info',
+    LOG_TO_CONSOLE: ENV.LOG_TO_CONSOLE !== 'false',
```
Comment (frbattid) [inline src/sth_configuration.js]: I understand this ('LOG_TO_CONSOLE`) and next (`LOG_TO_FILE`) are the default values... If both of them are false, that means by default nothing is logged? Maybe one of them (probably console) should be at least true... what do you think?

File: src/sth_logger.js [hunk 4]
```diff
@@ -3,21 +3,101 @@
...
+      ((log.level === LOG_LEVEL.fatal) ? ', alarm_status=ALARM' : '');
+  }
+
+  module.exports = function(theSthConfig) {
+    sthConfig = theSthConfig;
+
+    // Create the directory for the logging files if it does not exist
+    if (fs.existsSync(sthConfig.LOG_DIR)) {
+      if (!fs.statSync(sthConfig.LOG_DIR).isDirectory()) {
+        fs.mkdirSync(sthConfig.LOG_DIR);
```
Comment (frbattid) [inline src/sth_logger.js]: Does it create the log dir in a recursive way? E.g. `foo/bar/` is created?

File: src/sth_app.js (hunk 1 above)
Comment (gtorodelvalle) [inline src/sth_app.js]: Yeap, absolutely right! The winston (Node.js) module allows passing a "metadata" (object) with additional information. I used this "metadata" to pass to the logger the context in which the reason of the logging took place, according to the FIWARE logging format ;)

File: src/sth_app.js (hunk 2 above)
Comment (gtorodelvalle) [inline src/sth_app.js]: Ups, I just noticed a typo there :D s/exis/exist ;)

Comment (gtorodelvalle) [inline src/sth_app.js]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/02fb9825d614d8937e22617710b8f94c2fef13dd

File: src/sth_configuration.js (hunk 3 above)
Comment (gtorodelvalle) [inline src/sth_configuration.js]: No, no, it is just the opposite, I mean if the environment variables LOG_TO_CONSOLE and LOG_TO_FILE are not equal to 'false' (this is, their values are not explicitly set to 'false') the logs will be sent to console and to the file set ;)

File: src/sth_app.js (hunk 1 above)
Comment (frbattid) [inline src/sth_app.js]: Understood.

NTC

File: src/sth_configuration.js (hunk 3 above)
Comment (frbattid) [inline src/sth_configuration.js]: You are right, I did not notice the negation character :S :)

NTC

File: src/sth_logger.js (hunk 4 above)
Comment (gtorodelvalle) [inline src/sth_logger.js]: Really nice catch! ;) It did not considered that case ;)

Fixed in https://github.com/telefonicaid/IoT-STH/commit/8402aea1a5cbd42d3f0f84038f1f4a824e7e7e30

**Comment by frbattid** [comment]
LGTM

**Pull request merged**